llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -2916,8 +2916,11 @@ void CombinerHelper::replaceInstWithFConstant(MachineInstr &MI, double C) {
 
 void CombinerHelper::replaceInstWithConstant(MachineInstr &MI, int64_t C) {
   assert(MI.getNumDefs() == 1 && "Expected only one def?");
-  Builder.buildConstant(MI.getOperand(0), C);
-  MI.eraseFromParent();
+  LLT DstTy = MRI.getType(MI.getOperand(0).getReg());
+  if (isConstantLegalOrBeforeLegalizer(DstTy)) {
```

Comment:
I guess I'd expect this to be checked in the match rule

Reply:
It isn't. See above.

Reply:
Just because a G_ZEXT is sitting on a register doesn't imply that a G_CONSTANT is legal on that register. GlobalISel is for me for weird hardware.

Reply:
It doesn't, but should it

Reply:
This sentence is wrong.

Reply:
With the current code, we can build illegal MIR after the legalizer.

Reply:
```
if (MRI.hasOneNonDbgUse(DefReg)) {
  Helper.replaceInstWithConstant(MI, NewConst);
  MI.eraseFromParent();
}
```

You see the problem?

Reply:
It depends on where this code is running.
```
void CombinerHelper::replaceInstWithConstant(MachineInstr &MI, int64_t C) {
  assert(MI.getNumDefs() == 1 && "Expected only one def?");
  
  LLT DstTy = MRI.getType(MI.getOperand(0).getReg());
  Register SrcReg = MI.getOperand(1).getReg();
  if (isConstantLegalOrBeforeLegalizer(DstTy) && MRI.hasOneNonDbgUse(SrcReg)) {
    Builder.buildConstant(MI.getOperand(0), C);
    MI.eraseFromParent();
  }
}
```

Reply:
No.

Reply:
We never check the `DefReg`.

Reply:
A function called `replaceInstWithConstant` intended to be a helper routine should do as it's told. You can see in the above example it's easy to assume that it does in fact replace the MI as requested and therefore we'd get bugs. It's about fixing problems in the most natural place.

> This sentence is wrong.

I don't know what you expected to achieve with this comment but it's not helpful. Matt was trying to say that this issue should be fixed in the matcher.

Reply:
I found `matchOperandIsUndef` being used three times in Combine.td. Once to combine into undef and twice to combine into zero. The task of a `matchX` function is to check for `X`.

It is a hard rule that we query the legalizer before building. Any discussions of deviations are a waste of time.

Reply:
There is a different kind of match functions:
```
matchFoo(MI, BuildFn);
```
Here, in fact `matchFoo` has to query the legalizer, because it builds something. If we use match functions as a test, they might not know what we intend to build.

Reply:
`matchAnyExplicitUseIsUndef` is the same. It is a test and used several times in Combine.td to combine into different results. It cannot check for legality because its used as a test or predicate.

Reply:
Ok, I misunderstood his comment. If we split combines into `match` and `apply`, `apply` is going to build something  and has to check for legality.

Reply:
You haven't addressed my problem about the API design.

Reply:
Do you want to change the name `replaceInstWithConstant` ? Or something else that I missed?

Reply:
When the `unary_undef_to_zero` combine hits, it replaces the `MI` with zero without checking for legality. It is forbidden. Where do you want be to put the legality check?

Reply:
Why is this suddenly a problem now? Didn't you only see this when looking at the other patch?

Maybe we should relax the assumption for constants.

Reply:
https://github.com/llvm/llvm-project/blob/19131c7f36e047898ea954ee5a187ac62f2ab09b/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L2911
https://github.com/llvm/llvm-project/blob/19131c7f36e047898ea954ee5a187ac62f2ab09b/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L2936

I noticed it in the other PR. I am strongly against a curated list of opcodes with weaker legalization constraints.

Reply:
The helper is probably not that useful if we have to care about legalization as is. Probably should move the predicate into all the matchers, rename the helper, and add an assertion that the constant result is legal/custom

Reply:
The legalization query should be close to the build to prevent bugs. We have many generic match functions that are used in many combines and don't have enough information to know what the result of the combine is going to be. In most cases the legalization query should be in the apply function with the build.

Reply:
> The legalization query should be close to the build to prevent bugs. We have many generic match functions that are used in many combines and don't have enough information to know what the result of the combine is going to be. In most cases the legalization query should be in the apply function with the build.

You can't just be skipping mutation in the apply functions. The whole point of the apply functions is to do a transform that the matchers have validated. That's how we designed the combiner from the start. You need to find another way to make this work.

Reply:
We explicitly split it into match and apply. The match function is matching a pattern on the MIR. If it found a pattern, it will return true. Then it is up to the apply function to mutate the IR. Splitting legalization  between match and apply or moving it into match is error prone. The next refactoring will be really bad. The match function is for higher level decisions. I found MIR that looks like sin. apply is for low-level stuff. Let's build it.

Reply:
Let's go one step back. This discussion is unrelated to this PR. What is it that you don't like about this PR:
* I fixed a bug. We build constants without checking for legality.
* The legality check is too close to the build and you would rather see it 100 miles afar from the build.
* Otherwise, I would need some help to understand the unhappiness.

Reply:
> Splitting legalization between match and apply or moving it into match is error prone. The next refactoring will be really bad. The match function is for higher level decisions.

You're assuming this is true when it's just your opinion.

Reply:
Unrelated post.

Reply:
I do not understand what is bad or difficult about having the matcher check the legality. The matcher always has to understand what the apply is going to do.

Reply:
> Furthermore, I prefer the legality check to be as close as possible to the build to prevent bugs.

I don't see how one can hold this preference and yet at the same time be completely ok with the fact that the place which is named specifically for the job this is trying to do, match patterns, is not the place where the code lives. Instead the check is being added to a function who's sole purpose is to execute a transformation which has already been checked for legality in other senses, such as correctness and code quality heuristics, but now can fall over at the last second because of a completely random check that's only there because someone thought it would prevent bugs.

Even if you thought the matcher was a sub-optimal place to put the check due to ease-of-bugs issues, the right approach to solving the problem would surely be to think of another design, perhaps extend the tablegen-based combiner infrastructure to allow expressing these predicates in more ergonomic ways? Or is this your approach because it's simply the quickest way to do it?

Reply:
The reason I'm being resistant to your patches in recent times is not because I'm trying to be difficult for the sake of it. My goal is to try to hold the line in quality and overall good engineering practices (with a healthy dose of pragmatism). For AArch64 this framework is supposed to be the replacement for SelectionDAG in the long term future and we may have the design being used to compile trillions of lines of code, for software that's used directly or indirectly by almost every person on earth. The weight of that responsibility means that I would rather we not fix issues/add features than to degrade the quality of our codebase with a fatal thousand cuts. A quick fix now that compromises design cleanliness could bite us 10x worse in the future.

And despite the above, I'm not opposed to fixing issues on an iterative basis. We don't have to reach the final destination in one go, but it would be good to have some sort of plan. E.g. "I know this isn't an ideal solution, but I think the trade offs are worth it, and I'm willing to rethink the design in future" is the mindset I want to cultivate.

Reply:
One option is to move away from the generic matchers `Helper.matchOperandIsUndef` to specific matchers for  `unary_undef_to_zero`, i.e., `Helper.matchUndefToZero`.


But I really really hate to have a function that builds something and I have to search around if and where the legality check is.

Reply:
> One option is to move away from the generic matchers `Helper.matchOperandIsUndef` to specific matchers for `unary_undef_to_zero`, i.e., `Helper.matchUndefToZero`.

Why not just check in the matcher C++ code? The matcher already knows what the intended output looks like, it just defers the actual mutation to the applicator.

> 
> But I really really hate to have a function that builds something and I have to search around if and where the legality check is.

If I understand right you're basically saying you don't like the combiner infrastructure as it was designed, because that's how it's supposed to work.

Reply:
I don't see a need for Helper.matchOperandIsUndef any longer. You can directly match a G_IMPLICIT_DEF input in tablegen. The match predicate code would then only need to check the legality of the new G_IMPLICIT_DEF or G_CONSTANT. 

The legality check would always be in the matcher, there shouldn't be any searching to do.

Reply:
I prefer my solution. The legality check and the build are close together and easy to reason about.

Firstly, Helper.matchOperandIsUndef does not know whether we are combining into undef or constants. How can it reason about legality of uncertainty? Secondly, legality checks and build are spread wide afar, which makes it hard to reason about correctness.

Reply:
Then the utility of matchOperandIsUndef is limited and should not be used, or supplemented with a legality check